src/viam/sdk/components/encoder/encoder.hpp

```diff
@@ -0,0 +1,114 @@
+/// @file components/encoder/encoder.hpp
+///
+/// @brief Defines a `Encoder` component.
+#pragma once
+
+#include <string>
+
+#include <viam/api/component/encoder/v1/encoder.pb.h>
+
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource_manager.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Encoder Classes related to the `Encoder` component.
+
+/// @class EncoderSubtype
+/// @brief Defines a `ResourceSubtype` for the `Encoder` component.
+/// @ingroup Encoder
+class EncoderSubtype : public ResourceSubtype {
+   public:
+    std::shared_ptr<ResourceServerBase> create_resource_server(
+        std::shared_ptr<ResourceManager> manager) override;
+    std::shared_ptr<ResourceBase> create_rpc_client(std::string name,
+                                                    std::shared_ptr<grpc::Channel> chan) override;
+    EncoderSubtype(const google::protobuf::ServiceDescriptor* service_descriptor)
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Encoder encoder.hpp "components/encoder/encoder.hpp"
+/// @brief An encoder is a device that is hooked up to motors to report a position
+/// @ingroup Encoder
+///
+/// This acts as an abstract base class to be inherited from by any drivers representing
+/// specific encoder implementations. This class cannot be used on its own.
+class Encoder : public ComponentBase {
+   public:
+    /// @enum position_type
+    enum position_type {
+        // Unspecified position type
+        UNSPECIFIED,
```

Comment:
Is it meaningful to reflect the UNSPECIFIED value of the underlying proto here? If the real values that are supported by the encoder are TICKS_COUNT and ANGLE_DEGREES, should it be possible in the C++ API for anything to meaningfully be UNSPECIFIED? Or should an error happen somewhere if UNSPECIFIED is obtained over the wire, and therefore the programmatic API has no need of UNSPECIFIED?

Reply:
Typescript SDK uses:
```
async getPosition(
    positionType: EncoderPositionType = EncoderPositionType.POSITION_TYPE_UNSPECIFIED,
    extra = {}
  ) {
```

In order to support additional types of encoders, I am leaning towards keeping UNSPECIFIED as otherwise we would have to throw an error for get_position() if it returned an UNSPECIFIED type. (which is a judgement call for whether or not that is good or bad behavior)